core/src/consensus.rs

```diff
@@ -418,9 +489,21 @@ impl Tower {
         epoch_vote_accounts: &HashMap<Pubkey, (u64, Account)>,
     ) -> SwitchForkDecision {
         let root = self.lockouts.root_slot.unwrap_or(0);
+
         self.last_voted_slot()
             .map(|last_voted_slot| {
-                let last_vote_ancestors = ancestors.get(&last_voted_slot).unwrap();
+                let last_vote_ancestors =
+                    ancestors.get(&last_voted_slot).unwrap_or_else(|| {
+                        if !self.is_stray_last_vote() {
+                            panic!("no ancestors found with slot: {}", last_voted_slot);
+                        } else {
+                            // bank_forks doesn't have corresponding data for the stray restored last vote,
+                            // meaning severe inconsistentcy between saved tower and ledger.
+                            // (corrupted or pruned ledger, or only saved tower is moved over to new setup?)
+                            panic!("Unable to get ancestors for stray last vote {:?}", self.stray_restored_slot());
```

Comment:
hmmmm this will panic if our last vote was on some fork A, and then we start from a snapshot of another fork B right? I.e. in this example: https://github.com/solana-labs/solana/pull/10718#discussion_r474468319

If the last vote was on slot `10`, then we start from a snapshot `2`, then the only keys in the `ancestor` map here will be `2` and `0`, so if you try to lookup `10` in `ancestors` here, this seems like it will panic.

Reply:
yeah, it looks like I was wrong again... I'll address.

Reply:
half done...: https://github.com/solana-labs/solana/pull/10718/commits/0c121a951f6dd5687b638d24b8ef8f56bff640a1

I'll comment later how hard to fix `test_no_optimistic_confirmation_violation_with_tower`...

Reply:
the promised detailed comment is here: https://github.com/solana-labs/solana/pull/10718#discussion_r490104064